### PR TITLE
Update EVO2 tests according to Hyena arch changes

### PR DESCRIPTION
**Pull request by farhadrgh**
### Description

https://github.com/NVIDIA/NeMo/pull/12856 introduces code reduction and perf improvements including standardizing input/output shapes for Hyena operators and consequentially reducing rearrangement overhead. This PR updates the EVO2 test to comply with those changes, 

### Type of changes

- [ ]  Bug fix (non-breaking change which fixes an issue)
- [ ]  New feature (non-breaking change which adds functionality)
- [ ]  Refactor
- [ ]  Documentation update
- [ ]  Other (please describe):

### CI Pipeline Configuration
Configure CI behavior by applying the relevant labels:

- [SKIP_CI](https://github.com/NVIDIA/bionemo-framework/blob/main/docs/docs/user-guide/contributing/contributing.md#skip_ci) - Skip all continuous integration tests
- [INCLUDE_NOTEBOOKS_TESTS](https://github.com/NVIDIA/bionemo-framework/blob/main/docs/docs/user-guide/contributing/contributing.md#include_notebooks_tests) - Execute notebook validation tests in pytest
- [INCLUDE_SLOW_TESTS](https://github.com/NVIDIA/bionemo-framework/blob/main/docs/docs/user-guide/contributing/contributing.md#include_slow_tests) - Execute tests labelled as slow in pytest for extensive testing

> [!NOTE]
> By default, the notebooks validation tests are skipped unless explicitly enabled.

#### Authorizing CI Runs

We use [copy-pr-bot](https://docs.gha-runners.nvidia.com/apps/copy-pr-bot/#automation) to manage authorization of CI
runs on NVIDIA's compute resources.

* If a pull request is opened by a trusted user and contains only trusted changes, the pull request's code will
  automatically be copied to a pull-request/ prefixed branch in the source repository (e.g. pull-request/123)
* If a pull request is opened by an untrusted user or contains untrusted changes, an NVIDIA org member must leave an
  `/ok to test` comment on the pull request to trigger CI. This will need to be done for each new commit.

### Usage

```python
TODO: Add code snippet
```

### Pre-submit Checklist

 - [ ] I have tested these changes locally
 - [ ] I have updated the documentation accordingly
 - [ ] I have added/updated tests as needed
 - [ ] All existing tests pass successfully

### CODE REVIEW

**Comment by cspades** [comment]
LGTM but will let John verify:
```
- features = rearrange(features, "l b d -> b l d").contiguous()
+ features = rearrange(features, "l b d -> b d l").contiguous()
```

**Review by cspades — APPROVED** [review]

File: sub-packages/bionemo-evo2/tests/bionemo/evo2/test_hyena_operators.py [hunk 1]
```diff
@@ -150,14 +148,14 @@ def test_gpu_forward(self, operator: ParallelShortHyenaOperator):
         g = operator.num_groups
         dg = operator.group_dim
 
-        x1 = torch.ones((batch_size, seq_len, g, dg), device=device)
-        x2 = torch.ones((batch_size, seq_len, g, dg), device=device)
-        v = torch.ones((batch_size, seq_len, g, dg), device=device)
+        x1 = torch.ones((batch_size, (g * dg), seq_len), device=device)
```
Comment (jstjohn) [inline sub-packages/bionemo-evo2/tests/bionemo/evo2/test_hyena_operators.py:151]: Is there a test somewhere covering that this still works with tensor parallel? It could be that moving sequence to the last dimension breaks tensor parallel because that has a lot of hardcoded assumptions about splitting on axis 1. Maybe if you run the brca notebook but with TP=2 (using the experimental bf16 model weights if doing this on a non fp8 node) and it still works, that would be good? Please post a manual verification to this effect.

**Review by jstjohn — APPROVED** [review]
Approved but see my comment in line about manual verification of tensor parallel correctness. Ideally the same could be done for CP=2, but I am not 100% that we have that working in the predict script.

File: sub-packages/bionemo-evo2/tests/bionemo/evo2/test_hyena_operators.py (hunk 1 above)
Comment (farhadrgh) [reply sub-packages/bionemo-evo2/tests/bionemo/evo2/test_hyena_operators.py:151]: I am not aware of any tests for TP. But all the tests in NeMo and BioNeMo are passing. The CI failure now is discussed in [this thread](https://nvidia.slack.com/archives/C074Z808N05/p1744054519827599) and is unrelated to these changes.

I will run the notebook with TP=2 and report the results here

Comment (farhadrgh) [reply sub-packages/bionemo-evo2/tests/bionemo/evo2/test_hyena_operators.py:151]: I can now confirm that the notebook is reproducing ToT results with TP=2 or CP=2 on two A6000. However, there is a regression in ToT compared to the last time notebook was executed and this is unrelated to changes here ([more info regarding ToT regression](https://nvidia.slack.com/archives/C085X43PQEM/p1744298958418539))

**Comment by farhadrgh** [comment]
Need to bump NeMo to get the changes in https://github.com/NVIDIA/NeMo/pull/12988 after its merged

**Comment by farhadrgh** [comment]
[6e23633](https://github.com/NVIDIA/bionemo-framework/pull/798/commits/6e23633dd3427c6a430445efb5f69379d1d240f6) changes from https://github.com/NVIDIA/bionemo-framework/pull/807

**Comment by farhadrgh** [comment]
/ok to test [ae21a03](https://github.com/NVIDIA/bionemo-framework/pull/798/commits/ae21a039ddd5a2c9d21d65c845893ad58171e2d9)

**Comment by farhadrgh** [comment]
/ok to test [dc69464](https://github.com/NVIDIA/bionemo-framework/pull/798/commits/dc6946403563eb6820f13e40cef0bec16fbac25c)

File: sub-packages/bionemo-llm/src/bionemo/llm/model/loss.py [hunk 2]
```diff
@@ -181,17 +177,7 @@ def forward(
...
-            loss_for_microbatch = masked_token_loss(unreduced_token_loss, batch["loss_mask"])
-        else:
-            # reduce the loss across the micro batch per valid token.
-            # TODO(@jomitchell): Figure out who defines "num_valid_tokens_in_ub" in the batch and document/understand this.
-            #  This has something to do with context parallel, and there is probably a megatron or nemo function that adds this and
-            #  other necessary keys to the batch. Thanks!
-            loss_for_microbatch = masked_token_loss_context_parallel(
-                unreduced_token_loss, batch["loss_mask"], batch["num_valid_tokens_in_ub"]
-            )
+        loss_for_microbatch = masked_token_loss(unreduced_token_loss, batch["loss_mask"], cp_size)
```
Comment (trvachov) [inline sub-packages/bionemo-llm/src/bionemo/llm/model/loss.py:180]: 👀 @cspades

File: sub-packages/bionemo-llm/src/bionemo/llm/model/loss.py [hunk 3]
```diff
@@ -19,11 +19,7 @@
 from megatron.core import parallel_state, tensor_parallel
 from megatron.core.fusions.fused_cross_entropy import fused_vocab_parallel_cross_entropy
 from nemo.collections.nlp.modules.common.megatron.utils import average_losses_across_data_parallel_group
-from nemo.lightning.megatron_parallel import (
-    MegatronLossReduction,
-    masked_token_loss,
-    masked_token_loss_context_parallel,
-)
+from nemo.lightning.megatron_parallel import MegatronLossReduction, masked_token_loss
```
Comment (trvachov) [inline sub-packages/bionemo-llm/src/bionemo/llm/model/loss.py:22]: @cspades 👀

File: sub-packages/bionemo-llm/src/bionemo/llm/model/loss.py (hunk 2 above)
Comment (trvachov) [reply sub-packages/bionemo-llm/src/bionemo/llm/model/loss.py:180]: Why was the if statement removed @farhadrgh ? This `masked_token_loss_context_parallel` probably reduces along context parallel group.

Comment (cspades) [reply sub-packages/bionemo-llm/src/bionemo/llm/model/loss.py]: Fixed and merged into this PR.
